Realm/RLMSyncPermissionResults.mm

```diff
@@ -0,0 +1,160 @@
+}
+
+- (RLMSyncPermissionValue *)objectAtIndex:(NSInteger)index {
+    REALM_ASSERT_DEBUG(_results);
+    try {
+        return [[RLMSyncPermissionValue alloc] initWithPermission:_results->get(index)];
+    } catch (std::exception const& ex) {
+        @throw RLMException(ex);
+    }
+}
+
+- (instancetype)initWithResults:(std::unique_ptr<PermissionResults>)results {
+    if (self = [super init]) {
+        REALM_ASSERT_DEBUG(results);
+        _results = std::move(results);
+    }
+    return self;
+}
+
+- (RLMSyncPermissionValue *)firstObject {
+    return self.count == 0 ? nil : [self objectAtIndex:0];
+}
+
+- (RLMSyncPermissionValue *)lastObject {
+    return self.count == 0 ? nil : [self objectAtIndex:(self.count - 1)];
+}
+
+- (NSInteger)indexOfObject:(RLMSyncPermissionValue *)object {
+    for (int i=0; i<self.count; i++) {
+        if ([[self objectAtIndex:i] isEqual:object]) {
+            return i;
+        }
+    }
+    return NSNotFound;
+}
+
+- (RLMSyncPermissionResults *)objectsWithPredicate:(NSPredicate *)predicate {
+    REALM_ASSERT_DEBUG(_results);
+    auto& results = _results->results();
+    auto query = RLMPredicateToQuery(predicate,
+                                     [RLMObjectSchema objectSchemaForObjectStoreSchema:results.get_object_schema()],
+                                     [RLMSchema dynamicSchemaFromObjectStoreSchema:results.get_realm()->schema()],
+                                     results.get_realm()->read_group());
+    auto filtered_results = std::make_unique<PermissionResults>(_results->filter(std::move(query)));
+    return [[RLMSyncPermissionResults alloc] initWithResults:std::move(filtered_results)];
+}
+
+- (RLMSyncPermissionResults *)sortedResultsUsingProperty:(RLMSyncPermissionResultsSortProperty)property
+                                               ascending:(BOOL)ascending {
+    REALM_ASSERT_DEBUG(_results);
+    auto& results = _results->results();
+    std::string property_name;
+    switch (property) {
+        case RLMSyncPermissionResultsSortPropertyPath:
+            property_name = "path";
+            break;
+        case RLMSyncPermissionResultsSortPropertyUserID:
+            property_name = "user_id";
+            break;
+    }
+    const auto& table = results.get_tableview().get_parent();
+    size_t col_idx = table.get_descriptor()->get_column_index(property_name);
+    REALM_ASSERT(col_idx != size_t(-1));
+    auto sorted_results = std::make_unique<PermissionResults>(_results->sort({
+        table, {{ col_idx }}, { static_cast<bool>(ascending) }
+    }));
+    return [[RLMSyncPermissionResults alloc] initWithResults:std::move(sorted_results)];
+}
+
+- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state
+                                  objects:(id __unsafe_unretained [])buffer
+                                    count:(NSUInteger)len {
+    NSUInteger thisSize = self.count;
```

Comment:
This fast enumeration logic is pretty hairy. Is there any value in sharing some of this code with the RLMArray/RLMResults/RLMCollection/RLMArrayLinkView implementations?